<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Контекст" data-toc-modified-id="Контекст-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Контекст</a></span></li><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-данных-и-осмотр" data-toc-modified-id="Загрузка-данных-и-осмотр-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Загрузка данных и осмотр</a></span></li><li><span><a href="#Анализ-данных" data-toc-modified-id="Анализ-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Анализ данных</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

<footer id="footer"></footer>

<h1 align='center'>SQL</h1>
<p align="center"><img src="https://i.ibb.co/c8Chgrz/SQL-Structured-Query-Language.jpg"></p>


## Контекст

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<div style="float:left;margin:0 10px 10px 0" markdown="1">
    <a href="#footer"><img src='https://img.shields.io/badge/К содержанию-&#x21A9-blue'></a>
</div>

___


## Импорт библиотек

In [ ]:
import pandas as pd
import seaborn as sns
import warnings
import numpy as np
from sqlalchemy import create_engine 

sns.set_style('darkgrid')
warnings.filterwarnings("ignore")
pd.set_option('chained_assignment', None)
np.random.seed(42)

Создадим подключение к `БД`:

In [ ]:
db_config = {'user': 'praktikum_student', 
            'pwd': 'Sdf4$2;d-d30pp', 
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, 
            'db': 'data-analyst-final-project-db'} 


connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string) 

In [ ]:
def select_sql(sql):
    """вспомогательная функция для запросов"""
    return pd.io.sql.read_sql(sql, con = engine)

<div style="float:left;margin:0 10px 10px 0" markdown="1">
    <a href="#footer"><img src='https://img.shields.io/badge/К содержанию-&#x21A9-blue'></a>
</div>

___


## Загрузка данных и осмотр

Посмотрим данные и ознакомимся с атрибутивным составом

In [ ]:
sql = '''SELECT *
FROM   books t
LIMIT  5 '''

In [ ]:
select_sql(sql)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [ ]:
sql = '''SELECT *
FROM   authors t
LIMIT  5 '''

In [ ]:
select_sql(sql)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
sql = '''SELECT *
FROM   publishers t
LIMIT  5 '''

In [ ]:
select_sql(sql)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
sql = '''SELECT *
FROM   ratings t
LIMIT  5 '''

In [ ]:
select_sql(sql)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [ ]:
sql = '''SELECT *
FROM   reviews t
LIMIT  5 '''

In [ ]:
select_sql(sql)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


Нам предоставлена информация о книгах. Ниже представлен атрибутивный состав данных и взаимосвязь сущностей

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<img src="https://i.ibb.co/ZgK9RmD/Untitled-33.png" alt="Untitled-33" border="0">


Нам необходимо произвести анализ базы данных и ответить на следующие вопросы:  

- Сколько книг вышло после 1 января 2000 года;  
- Для каждой книги посчитать количество обзоров и среднюю оценку;  
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок.  

<div style="float:left;margin:0 10px 10px 0" markdown="1">
    <a href="#footer"><img src='https://img.shields.io/badge/К содержанию-&#x21A9-blue'></a>
</div>

___


## Анализ данных

In [ ]:
sql = '''SELECT *
FROM   books t
LIMIT  5 '''

In [ ]:
select_sql(sql)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Посмотрим сколько книг вышло, после 1 января 2000 года

In [ ]:
sql = '''
SELECT 
    COUNT(DISTINCT(book_id)) as quantity
FROM 
    books t
WHERE 
    t.publication_date > '2000-01-01'
'''

In [ ]:
select_sql(sql)

quantity
0       821

После 1 января 2020 года было выпущено 821 уникальных книг. Посмотрим среднюю оценку и количетсво отзывов

In [ ]:
sql = '''
WITH review AS
(
         SELECT   Count (DISTINCT(review_id)) AS review_cnt,
                  book_id
         FROM     reviews
         GROUP BY book_id), rating AS
(
         SELECT   Avg(rating :: real) AS avg_rating,
                  book_id
         FROM     ratings
         GROUP BY book_id)
SELECT    t.book_id,
          t.title,
          CASE
                    WHEN r.review_cnt IS NULL THEN 0
                    ELSE r.review_cnt
          END AS cnt,
          CASE
                    WHEN ra.avg_rating::real IS NULL THEN 0
                    ELSE ra.avg_rating::real
          END AS rating
FROM      books t
LEFT JOIN review r
ON        t.book_id = r.book_id
LEFT JOIN rating ra
ON        r.book_id = ra.book_id
ORDER BY  cnt DESC limit 15'''

In [ ]:
select_sql(sql)

book_id                                              title  cnt   rating
0       948                            Twilight (Twilight  #1)    7  3.66250
1       854                                           The Road    6  3.77273
2       207                                    Eat  Pray  Love    6  3.39583
3       656                                     The Book Thief    6  4.26415
4       733                          The Giver (The Giver  #1)    6  3.75000
5       779  The Lightning Thief (Percy Jackson and the Oly...    6  4.08065
6       302  Harry Potter and the Prisoner of Azkaban (Harr...    6  4.41463
7       299  Harry Potter and the Chamber of Secrets (Harry...    6  4.28750
8       497                          Outlander (Outlander  #1)    6  4.12500
9       627                                      The Alchemist    6  3.78947
10      695  The Curious Incident of the Dog in the Night-Time    6  4.08108
11      696             The Da Vinci Code (Robert Langdon  #2)    6  3.83051
12      734                                   The Glass Castle    6  4.20690
13      750                The Hobbit  or There and Back Again    6  4.12500
14      673                             The Catcher in the Rye    6  3.82558

По количеству отзывов лидируют `Сумерки`. Но рейтинг не самый большой. Посмотрим, кто лидирует по оценкам. Также в топ-15 присутствуют книги из серии о `Гарри Поттере`, `Хоббит`

In [ ]:
sql = '''
WITH review AS
(
         SELECT   Count (DISTINCT(review_id)) AS review_cnt,
                  book_id
         FROM     reviews
         GROUP BY book_id), rating AS
(
         SELECT   Avg(rating :: real) AS avg_rating,
                  book_id
         FROM     ratings
         GROUP BY book_id)
SELECT    t.book_id,
          t.title,
          CASE
                    WHEN r.review_cnt IS NULL THEN 0
                    ELSE r.review_cnt
          END AS cnt,
          CASE
                    WHEN ra.avg_rating::real IS NULL THEN 0
                    ELSE ra.avg_rating::real
          END AS rating
FROM      books t
LEFT JOIN review r
ON        t.book_id = r.book_id
LEFT JOIN rating ra
ON        r.book_id = ra.book_id
ORDER BY  rating DESC  limit 15'''

In [ ]:
select_sql (sql)

book_id                                              title  cnt  rating
0       182                                         Dead Souls    2     5.0
1       275                                          Geek Love    2     5.0
2       169  Crucial Conversations: Tools for Talking When ...    2     5.0
3        86      Arrows of the Queen (Heralds of Valdemar  #1)    2     5.0
4       229           Evil Under the Sun (Hercule Poirot  #24)    2     5.0
5       224                                      Evening Class    2     5.0
6        57                    Act of Treason (Mitch Rapp  #9)    2     5.0
7        20              A Fistful of Charms (The Hollows  #4)    2     5.0
8        55          A Woman of Substance (Emma Harte Saga #1)    2     5.0
9        76                                        Angels Fall    2     5.0
10       62                                      Alas  Babylon    2     5.0
11      213                      Emily of New Moon (Emily  #1)    2     5.0
12      136  Captivating: Unveiling the Mystery of a Woman'...    2     5.0
13       17                      A Dirty Job (Grim Reaper  #1)    4     5.0
14      297                                         Hard Times    2     5.0

По оценкам, лидирует классика - `Мертвые души`, но при этом - у книги мало отзывов. Чаще всего читают более легкие книги. Посмотрим на издательство, которое выпустило наибольшее количество книг толще 50 страниц

In [ ]:
sql = '''
WITH books_50 AS
(
         SELECT   Count(DISTINCT(book_id)) AS book_cnt,
                  t.publisher_id
         FROM     books t
         WHERE    t.num_pages > 50
         GROUP BY publisher_id)
SELECT    p.publisher as best_publisher,
          CASE
                    WHEN b.book_cnt IS NULL THEN 0
                    ELSE b.book_cnt
          END AS books_cnt
FROM      publishers p
LEFT JOIN books_50 b
ON        p.publisher_id = b.publisher_id
ORDER BY  books_cnt DESC limit 1
'''

In [ ]:
select_sql(sql)

best_publisher  books_cnt
0  Penguin Books         42

`Penguin Books` выпустило наибольшее количество книг. Количество книг - `42`. Посмотрим, кто пишет самые обсуждаемые книги и найдем среднюю оценку автора

In [ ]:
sql = '''
WITH rating_50 AS
(
         SELECT   Count(rating_id)     AS cnt,
                  Avg(rating::real) AS avg_rating,
                  book_id
         FROM     ratings
         GROUP BY book_id
         HAVING   Count(rating_id) >= 50)
SELECT    a.author,
          AVG(r.avg_rating::real) as rating
FROM      books t
INNER JOIN rating_50 r
ON        r.book_id = t.book_id
INNER JOIN authors a
ON        a.author_id = t.author_id

GROUP BY
    a.author_id,
    a.author
ORDER BY rating DESC
LIMIT 1
'''

In [ ]:
select_sql(sql)

author    rating
0  J.K. Rowling/Mary GrandPré  4.283844

Самый популярный и обсуждаемый автор - `J.K. Rowling` - автор книг о Гарри Поттере. Посмотрим среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
sql ='''
WITH selected_50
     AS (SELECT username,
                Count(*) AS reviews_num
         FROM   reviews
         WHERE  username IN (SELECT username
                             FROM   ratings
                             GROUP  BY username
                             HAVING Count(rating_id) > 50)
         GROUP  BY username)
SELECT Avg(reviews_num)
FROM   selected_50 

'''

In [ ]:
select_sql(sql)

avg
0  24.333333

Пользователи, поставившие больше `50` оценок - в среднем написали `24` обзора

<div style="float:left;margin:0 10px 10px 0" markdown="1">
    <a href="#footer"><img src='https://img.shields.io/badge/К содержанию-&#x21A9-blue'></a>
</div>

___


## Выводы

Мы проанализировали предоставленную базу данных и ответили на поставленные вопросы. Резюмируя:
    
    1. После 1 января 2020 года было выпущено 821 уникальных книг. 
    2. По количеству отзывов лидируют Сумерки, в рейтинге присутствует серия книг о Гарри Поттере
    3. По оценкам, лидирует классика - Мертвые души.
    4. Penguin Books выпустило наибольшее количество книг. Количество книг - 42.
    5. Самый популярный и обсуждаемый автор - J.K. Rowling - автор книг о Гарри Поттере.
    6. Пользователи, поставившие больше `50` оценок - в среднем написали `24` обзора
    
Таким образом, на момент исследования, стоит обратить внимание на многосерийные книги, приключенческого характера

<div style="float:left;margin:0 10px 10px 0" markdown="1">
    <a href="#footer"><img src='https://img.shields.io/badge/К содержанию-&#x21A9-blue'></a>
</div>

___
